### PR TITLE
Automated cherry pick of #114998

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

#115324 and #114998 Made OpenAPIV3 schemas an explicit requirement to start the APIServer for performance and to fix a bug where previously server-side-apply would be silently disabled without OpenAPI config. 

Unfortunately #114998 did not make it into 1.27, and fixes a bug given by 1.27, and throws a more explicit error message rather than failing silently.

#### Which issue(s) this PR fixes:

Consumers of kube-apiserver may face issues upgrading without this PR which:

1. Gives a clearer error message about requirement of OpenAPIV3
2. Fixes bug where OpenAPIV2 config was used instead of V3

#### Does this PR introduce a user-facing change?

```release-note
Fixes bug where OpenAPIV2 config was used instead of V3, and gives clear error message about OpenAPIV3 requirement
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
/retest

(flake fixed in https://github.com/kubernetes/kubernetes/pull/117539 but not backported to 1.27)

**Review by apelisse — COMMENTED** [review]
Thanks
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 1]
```diff
@@ -953,13 +945,13 @@ func NewDefaultAPIGroupInfo(group string, scheme *runtime.Scheme, parameterCodec
...
 	if s.openAPIV3Config == nil {
-		//!TODO: A future work should add a requirement that
-		// OpenAPIV3 config is required. May require some refactoring of tests.
-		return nil, nil
+		// SSA is GA and requires OpenAPI config to be set
+		// to create models.
+		return nil, errors.New("OpenAPIV3 config must not be nil")
 	}
-	pathsToIgnore := openapiutil.NewTrie(s.openAPIConfig.IgnorePrefixes)
+	pathsToIgnore := openapiutil.NewTrie(s.openAPIV3Config.IgnorePrefixes)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:954]: do we ever actually put a value in IgnorePrefixes, or is this always `[]string{}` (regardless of whether we used `openAPIV3Config` or `openAPIConfig`?

**Comment by liggitt** [comment]
> Fixes bug where OpenAPIV2 config was used instead of V3

do you have a pointer to where that was happening, and an example of the user-facing symptom that would cause?

I don't see any change in the captured `api/openapi-spec/swagger.json` or `api/openapi-spec/v3/...` files, so it's not clear what user-facing change this is driving

**Comment by alexzielenski** [comment]
> > Fixes bug where OpenAPIV2 config was used instead of V3
> 
> do you have a pointer to where that was happening, and an example of the user-facing symptom that would cause?
> 
> I don't see any change in the captured `api/openapi-spec/swagger.json` or `api/openapi-spec/v3/...` files, so it's not clear what user-facing change this is driving

`IgnorePrefixes` is not used by us anywhere, but it can be set by downstream clients. I don't think it is likely to cause problems since I would expect the v2 and v3 list to be the same but some community members have flagged it:

https://github.com/kubernetes/kubernetes/pull/115324#discussion_r1323307399
https://github.com/kubernetes/kubernetes/pull/114998#issuecomment-1650928263

**Comment by liggitt** [comment]
> > > Fixes bug where OpenAPIV2 config was used instead of V3
> > 
> > 
> > do you have a pointer to where that was happening, and an example of the user-facing symptom that would cause?
> > I don't see any change in the captured `api/openapi-spec/swagger.json` or `api/openapi-spec/v3/...` files, so it's not clear what user-facing change this is driving
> 
> `IgnorePrefixes` is not used by us anywhere, but it can be set by downstream clients. I don't think it is likely to cause problems since I would expect the v2 and v3 list to be the same. Some community members have flagged it as a potential bug but I haven't seen any reports of anyone being bit by it:
> 
> [#115324 (comment)](https://github.com/kubernetes/kubernetes/pull/115324#discussion_r1323307399) [#114998 (comment)](https://github.com/kubernetes/kubernetes/pull/114998#issuecomment-1650928263) (OpenAPIV3 was made GA in 1.27 and this user misattributed newly-enabled-by-default `OpenAPIV3`'s error about lack of OpenAPIV3 config to the `IgnorePrefixes` line)

Ok. This is a ~big change that removes a lot of NPE checks/protections, which makes mis-wiring more obvious, but also exposes us to issues if construction of field managers / openapi is ever nil when we don't expect it to be. Are we 100% sure that all custom resources can produce a field manager, no matter what their schema is (even old v1beta1 ones that don't have structural schemas)?

**Comment by alexzielenski** [comment]
>  Are we 100% sure that all custom resources can produce a field manager, no matter what their schema is

Since #112748, a field manager is required to create a request scope for CRDs ([source](https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L879-L887), [source](https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L980-L996)). If the request scope or field manager cannot be created, the endpoint is not served already ([source](https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L297-L309)), so this PR does not affect the request path of CRDs for which a field manager cannot be constructed.

> (even old v1beta1 ones that don't have structural schemas)?

CRDs without schemas will use the `DeducedTypeConverter` in their field maanger. CRDs with invalid schemas are not served and are also unaffected. ([source](https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L678-L690))

**Comment by liggitt** [comment]
> CRDs with invalid schemas are not served and are also unaffected.

we still serve the CR, are you saying we don't serve the openapi?

**Comment by alexzielenski** [comment]
> > CRDs with invalid schemas are not served and are also unaffected.
> 
> we still serve the CR, are you saying we don't serve the openapi?

I mean the `Apply` and `Update` and update handlers are not available. In CRD handler's `ServeHTTP`:

https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L297-L309

if `getOrCreateServingInfo` returns an error, an error is returned for the request and `serveResource` which calls the apply/update handlers is unreachable.

**Comment by alexzielenski** [comment]
I see #82438 was meant to use the deduced converter when models could not be build. In https://github.com/kubernetes/kubernetes/pull/95736/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eR672-R679 an early return was added to CRDs for which schema isn't convertible to SMD schema

**Comment by liggitt** [comment]
> I see #82438 was meant to use the deduced converter when models could not be built. In https://github.com/kubernetes/kubernetes/pull/95736/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eR672-R679 an early return was added to CRDs for which schema isn't convertible to SMD schema. Seems like a bug?

if `buildOpenAPIModelsForApply` returns an error for a given CRD, it still looks like we null out the schema:

```go
	openAPIModels, err := buildOpenAPIModelsForApply(r.staticOpenAPISpec, crd)
	if err != nil {
		utilruntime.HandleError(fmt.Errorf("error building openapi models for %s: %v", crd.Name, err))
		openAPIModels = nil
	}
```

and then use the deduced type converter (since openAPIModels is nil, we don't get to the managedfields.NewTypeConverter call):
```go
	var typeConverter managedfields.TypeConverter = managedfields.NewDeducedTypeConverter()
	if len(openAPIModels) > 0 {
		typeConverter, err = managedfields.NewTypeConverter(openAPIModels, crd.Spec.PreserveUnknownFields)
		if err != nil {
			return nil, err
		}
	}
```

that looks like we will always have a non-nil type converter, which is what we want

**Comment by liggitt** [comment]
is there any case where `buildOpenAPIModelsForApply` would return successfully and `managedfields.NewTypeConverter` would fail?

**Comment by alexzielenski** [comment]
> that looks like we will always have a non-nil type converter, which is what we want

Ah #82438 initially seemed like the intention was maybe to use the deduced type converter in cases like this, but I see now it was intended for CRD schemas not expressible in OpenAPI, which the current implementation does correctly. 

> is there any case where buildOpenAPIModelsForApply would return successfully and managedfields.NewTypeConverter would fail?

`managedfields.NewTypeConverter` can only fail if `ToSchemaFromOpenAPI` fails:

https://github.com/kubernetes/kubernetes/blob/063fd3b8bc4ef5ccc2fa561624f22d5178160c69/staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/typeconverter.go#L45-L55

The possible error returns are:

- [x] [Invalid `type`](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L118-L124)
    Checked in CRD Validation [source](https://github.com/kubernetes/kubernetes/blob/4c43a25e6a430f686998f6ef25bbc79b0fa1d6a5/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1308-L1310)
- [x] Invalid `x-kubernetes-map-type` [[1]](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L213-L216)[[2]](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L141-L144)
    Checked in CRD Validation [source](https://github.com/kubernetes/kubernetes/blob/4c43a25e6a430f686998f6ef25bbc79b0fa1d6a5/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L992-L994)
- [x] [Invalid `x-kubernetes-list-type`](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L226-L229)
    Checked in CRD Validation [source](https://github.com/kubernetes/kubernetes/blob/4c43a25e6a430f686998f6ef25bbc79b0fa1d6a5/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L996-L1001)
- [ ] [`items` can only have one subitem](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L232-L233)
    Not Checked in CRD Validation. 
    ToSchemaFromOpenAPI was added in 1.27 to replace slower `ToProtoModels` and mirror the same behavior. This error [exists in ToProtoModels](https://github.com/kubernetes/kube-openapi/blob/9b4dcd38a4bf055aff3ef4ab23eeadc84ad1361b/pkg/util/proto/document.go#L214-L2180) and has always been thrown (erroneously) by SSA. When SSA was enabled by default this would have surfaced as a regression in CRDs not being handled if multiple `items` is used. (Unless there is another place that enforces that requirement I'm missing)
- [x] [Missing array `items`](https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/schemaconv/openapi.go#L244-L246)
    Checked in CRD Validation [source](https://github.com/kubernetes/kubernetes/blob/4c43a25e6a430f686998f6ef25bbc79b0fa1d6a5/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1348-L1350)

All except one of these are be checked in the structural schema validator.

**Comment by liggitt** [comment]
thanks for tracing in https://github.com/kubernetes/kubernetes/pull/120612#issuecomment-1722280246 ... that level of checking is needed whenever we're backporting stuff, unfortunately, and anything we can do in `master` to make some of that more obviously equivalent and safe would be welcome

/lgtm
/approve
/cc @kubernetes/release-managers

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**